server/rating_service/rating_service.py

```diff
@@ -0,0 +1,382 @@
+
+    async def _persist_rating_changes(
+        self,
+        game_id: int,
+        rating_type: RatingType,
+        old_ratings: Dict[PlayerID, Rating],
+        new_ratings: Dict[PlayerID, Rating],
+        outcomes: Dict[PlayerID, GameOutcome],
+    ) -> None:
+        """
+        Persist computed ratings to the respective players' selected rating
+        """
+        self._logger.info("Saving rating change stats for game %i", game_id)
+
+        async with self._db.acquire() as conn:
+            for player_id, new_rating in new_ratings.items():
+                old_rating = old_ratings[player_id]
+                self._logger.debug(
+                    "New %s rating for player with id %s: %s -> %s",
+                    rating_type,
+                    player_id,
+                    old_rating,
+                    new_rating,
+                )
+                await conn.execute(
+                    "UPDATE game_player_stats "
+                    "SET after_mean = %s, after_deviation = %s, "
+                    "mean = %s, deviation = %s, scoreTime = NOW() "
+                    "WHERE gameId = %s AND playerId = %s",
+                    (
+                        new_rating.mu,
+                        new_rating.sigma,
+                        old_rating.mu,
+                        old_rating.sigma,
+                        game_id,
+                        player_id,
+                    ),
+                )
+
+                gps_rows = await conn.execute(
+                    select([gps_table.c.id]).where(
+                        and_(
+                            gps_table.c.playerId == player_id,
+                            gps_table.c.gameId == game_id,
+                        )
+                    )
+                )
+                gps_row = await gps_rows.fetchone()
+                if gps_row is None:
+                    self._logger.warning(
+                        f"No game_player_stats entry for player {player_id} of game {game_id}."
+                    )
+                    raise EntryNotFoundError
+                game_player_stats_id = gps_row[gps_table.c.id]
+
+                await self._update_rating_tables(
+                    conn,
+                    game_player_stats_id,
+                    rating_type,
+                    player_id,
+                    new_rating,
+                    old_rating,
+                    outcomes[player_id],
+                )
+
+                self._update_player_object(player_id, rating_type, new_rating)
+
+    async def _update_rating_tables(
+        self,
+        conn,
+        game_player_stats_id: int,
+        rating_type: RatingType,
+        player_id: PlayerID,
+        new_rating: Rating,
+        old_rating: Rating,
+        outcome: GameOutcome,
+    ) -> None:
+
+        is_victory = outcome is GameOutcome.VICTORY
+        rating_type_id = self._rating_type_ids[rating_type.value]
+
+        await conn.execute(
```

Comment:
Insert expressions are actually really slick in SQLAlchemy: https://docs.sqlalchemy.org/en/13/core/tutorial.html#insert-expressions